app/src/main/java/de/westnordost/streetcomplete/quests/smoothness/AddSmoothnessForm.kt

```diff
@@ -0,0 +1,166 @@
+            .setPositiveButton(android.R.string.ok, null)
+            .show()
+        }
+    }
+
+    // this requires Smoothness class to be sorted from best to worst!
+    private fun getAnswers(minSmoothness: Int, maxSmoothness: Int): List<Item<Smoothness>> {
+        val answers = mutableListOf<Item<Smoothness>>()
+        for (smoothness in Smoothness.values().sliceArray(minSmoothness..maxSmoothness))
+            answers.add(Item(smoothness, getImage(smoothness), smoothness.title, getDescription(smoothness)))
+        return answers
+    }
+
+    private fun getDescription(smoothness: Smoothness) = when (smoothness) {
+        Smoothness.EXCELLENT -> when (surfaceTag) {
+            "paving_stones" -> R.string.quest_smoothness_description_excellent_paving_stones
+            else -> R.string.quest_smoothness_description_excellent
+        }
+        Smoothness.GOOD -> when (surfaceTag) {
+            "paving_stones" -> R.string.quest_smoothness_description_good_paving_stones
+            else -> R.string.quest_smoothness_description_good
+        }
+        Smoothness.INTERMEDIATE -> when (surfaceTag) {
+            "paving_stones", "sett" -> R.string.quest_smoothness_description_intermediate_paving_stones
+            "compacted" -> R.string.quest_smoothness_description_intermediate_compacted
+            else -> when (highwayTag) {
+                in ALL_PATHS_EXCEPT_STEPS -> R.string.quest_smoothness_description_intermediate_path
+                else -> R.string.quest_smoothness_description_intermediate
+            }
+        }
+        Smoothness.BAD -> when (surfaceTag) {
+            "sett" -> R.string.quest_smoothness_description_bad_sett
+            "paving_stones" -> R.string.quest_smoothness_description_bad_paving_stones
+            else -> when (highwayTag) {
+                in ALL_PATHS_EXCEPT_STEPS -> R.string.quest_smoothness_description_bad_path
+                else -> R.string.quest_smoothness_description_bad_road
+            }
+        }
+        Smoothness.VERY_BAD -> when (highwayTag) {
+            in ALL_PATHS_EXCEPT_STEPS -> R.string.quest_smoothness_description_very_bad_path
+            else -> R.string.quest_smoothness_description_very_bad_road
+        }
+        Smoothness.HORRIBLE -> when (highwayTag) {
+            in ALL_PATHS_EXCEPT_STEPS -> R.string.quest_smoothness_description_horrible
+            else -> R.string.quest_smoothness_description_horrible
+        }
+        Smoothness.VERY_HORRIBLE -> when (highwayTag) {
+            in ALL_PATHS_EXCEPT_STEPS -> R.string.quest_smoothness_description_very_horrible
+            else -> R.string.quest_smoothness_description_very_horrible
+        }
+        Smoothness.IMPASSABLE -> when (highwayTag) {
+            in ALL_PATHS_EXCEPT_STEPS -> R.string.quest_smoothness_description_impassable
+            else -> R.string.quest_smoothness_description_impassable
+        }
+    }
+
+    private fun getImage(smoothness: Smoothness) = when (smoothness) {
```

Comment:
Since the image should be completely disjunct for the different surfaces, I think it is better to have one `getImageResId` function for every different surface type.

Additionally, this function could return null for when there is no image for a certain combination of smoothness+surface, i.e. it is a not a selectable choice. E.g. "good sett".
With this in place, the code in lines 29-35 are not necessary because simply those items could be returned that have an image and a description. (And if no items are returned, this would be a case of throwing an Exception)

Reply:
done